plugins/hls-refactor-plugin/src/Development/IDE/Plugin/CodeAction/Extract.hs

```diff
@@ -0,0 +1,268 @@
+      source'' <- addNewDef $ makeDeltaAst source'
+      let diff = makeDiffTextEdit (T.pack $ exactPrint parsedSource) (T.pack $ exactPrint $ makeDeltaAst source'')
+      pure [("Extract function", fromLspList diff)]
+  where
+    minimalRange = removeWhitespacePadding range srcText
+    newDefName = noLocA $ mkRdrUnqual $ mkVarOcc $ T.unpack "newDefinition"
+    execTransform = (\(a,_,_) -> a) . runTransform
+suggestExtractFunction _ _ _ = [] -- Could not find Annotated ParsedSource
+
+-- | This function represents a simple guess of which variables are going to be out of scope after the expression
+-- is extracted.  This strategy will probably work for >90% cases.
+--
+-- The strategy is that we assume that all variables bound in the extracted expr scope over any occurrence of said
+-- variable in the same extracted expr. Assuming this, we can take the intersection of the variables bound in the
+-- expression's scope, and those bound in the surrounding context, and thereby find the newly-free variables
+-- post-extraction.
+--
+-- A simple counter-example to this strategy would occur by extracting the rhs of `foo`:
+--    foo bar = let tmp = bar; bar = 1 in bar
+-- Where bar is used in the rhs of, but is also bound in the same expression. Thankfully this would just cause an
+-- unbound variable error post-extraction.
+--
+-- In order to have a more robust strategy without excessive maintenance burden, we would require output from
+-- GHC's renamer phase, preferably in-tree.
+--
+-- Some known weaknesses:
+-- TODO Some subtle name shadowing issues in the extracted expression. Hard to fix without more GHC help.
+--      In this case, bar is actually a free variable, but this function will find it not to be free.
+-- TODO Record `{..}` pattern matching syntax; we can find the fields in the matching record to fix this. See
+--      hls-explicit-refcord-fields-plugin
+findFreeVarsPostExtract :: (Monad m) => SrcSpan -> LHsExpr GhcPs -> ParsedSource -> TransformT m [RdrName]
+findFreeVarsPostExtract extractSpan extractExpr parsedSource = do
+      -- Find the variables that are bound in this declaration. Not finding the surrounding decl is a bug.
+      -- TODO return error when we cannot find the containing decl
+      (fromMaybe [] -> boundVarsInWholeDecl) <- querySmallestDeclWithM (pure . (extractSpan `isSubspanOf`)) (pure . boundVarsQ) parsedSource
+      let boundVarsInExtractedExpr = boundVarsQ extractExpr
+      let usedVarsInExtractedExpr = usedVarsQ extractExpr
+      let varsThatScopeOverExtractedExpr =
+            Set.fromList boundVarsInWholeDecl `Set.difference` Set.fromList boundVarsInExtractedExpr
+      -- We use intersection, which implicitly assumes that all non-bound variables in the surrounding scope are
+      -- imports. This is where problems with record field wildcards arise.
+      let postExtractFreeVars = varsThatScopeOverExtractedExpr `Set.intersection` Set.fromList usedVarsInExtractedExpr
```

Comment:
Hmm, now I'm a bit confused. Why is the calculation that we want not just `varsUsedInExpr - varsBoundInExpr`?